File: backend/app/services/nasabah_service.py
import uuid
from typing import Any, Dict, List, Optional, Tuple
from fastapi import HTTPException, status
from app.core.database import db
from app.core.security import get_password_hash
from app.schemas.nasabah import (
    NasabahCreateRequest,
    NasabahResponse,
    NasabahStatusUpdateRequest,
    NasabahUpdateRequest,
)


class NasabahService:
    @staticmethod
    def generate_next_account_number(conn) -> str:
        """Atomic account sequence increment returning bsuvh0000 to bsuvh9999."""
        cursor = conn.cursor()
        cursor.execute("SELECT last_number FROM account_sequences WHERE id = 1")
        row = cursor.fetchone()
        if not row:
            cursor.execute("INSERT INTO account_sequences (id, last_number) VALUES (1, -1)")
            last_number = -1
        else:
            last_number = row[0] if isinstance(row, (tuple, list)) else row["last_number"]
        
        next_number = last_number + 1
        if next_number > 9999:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Kapasitas nomor rekening bsuvh0000-bsuvh9999 telah habis. Silakan hubungi pengembang sistem.",
            )
        
        cursor.execute("UPDATE account_sequences SET last_number = ? WHERE id = 1", (next_number,))
        return f"bsuvh{next_number:04d}"

    @staticmethod
    def get_nasabah_balance(nasabah_id: str) -> int:
        """Calculate or fetch the latest balance of a nasabah."""
        latest_trx = db.fetch_one(
            """
            SELECT balance_after FROM transactions
            WHERE nasabah_id = ?
            ORDER BY transaction_date DESC, created_at DESC
            LIMIT 1
            """,
            (nasabah_id,)
        )
        if latest_trx:
            return latest_trx["balance_after"]
        return 0

    @classmethod
    def create_nasabah(
        cls,
        data: NasabahCreateRequest,
        registration_source: str = "SELF",
        creator_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create new nasabah and associated user credentials."""
        # 1. Check if NIK already exists
        existing_nik = db.fetch_one("SELECT id FROM nasabah WHERE nik = ?", (data.nik,))
        if existing_nik:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="NIK sudah terdaftar dalam sistem.",
            )

        nasabah_id = str(uuid.uuid4())
        user_id = str(uuid.uuid4())
        hashed_password = get_password_hash(data.password)

        with db.connection() as conn:
            # Generate account number
            account_no = cls.generate_next_account_number(conn)
            
            # Insert nasabah record
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT INTO nasabah (
                    id, customer_id, account_no, nik, name, phone, address,
                    rt, rw, kelurahan, kecamatan, kabupaten_kota, nasabah_category, email,
                    status, registration_source, created_by, updated_by
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'ACTIVE', ?, ?, ?)
                """,
                (
                    nasabah_id,
                    account_no,
                    account_no,
                    data.nik,
                    data.name,
                    data.phone,
                    data.address,
                    getattr(data, "rt", None),
                    getattr(data, "rw", None),
                    getattr(data, "kelurahan", None),
                    getattr(data, "kecamatan", None),
                    getattr(data, "kabupaten_kota", None),
                    getattr(data, "nasabah_category", None) or "Rumah Tangga/Individu",
                    data.email,
                    registration_source,
                    creator_id or "SYSTEM",
                    creator_id or "SYSTEM"
                )
            )

            # Insert login user record
            cursor.execute(
                """
                INSERT INTO users (
                    id, username, name, email, phone, password_hash, role, nasabah_id, status, is_active, created_by, updated_by
                ) VALUES (?, ?, ?, ?, ?, ?, 'NASABAH', ?, 'ACTIVE', 1, ?, ?)
                """,
                (user_id, account_no, data.name, data.email, data.phone, hashed_password, nasabah_id, creator_id or "SYSTEM", creator_id or "SYSTEM")
            )

        nasabah = db.fetch_one("SELECT * FROM nasabah WHERE id = ?", (nasabah_id,))
        nasabah["balance"] = 0
        return nasabah

    @classmethod
    def list_nasabah(
        cls,
        search: Optional[str] = None,
        status_filter: Optional[str] = None,
        page: int = 1,
        page_size: int = 20
    ) -> Tuple[List[Dict[str, Any]], int]:
        """List nasabah with search, filter, and pagination."""
        query = "SELECT * FROM nasabah WHERE 1=1"
        count_query = "SELECT COUNT(*) as total FROM nasabah WHERE 1=1"
        params: List[Any] = []
        count_params: List[Any] = []

        if search:
            search_pattern = f"%{search.strip()}%"
            filter_clause = " AND (name LIKE ? OR nik LIKE ? OR customer_id LIKE ? OR phone LIKE ? OR kelurahan LIKE ? OR kecamatan LIKE ? OR nasabah_category LIKE ?)"
            query += filter_clause
            count_query += filter_clause
            params.extend([search_pattern, search_pattern, search_pattern, search_pattern, search_pattern, search_pattern, search_pattern])
            count_params.extend([search_pattern, search_pattern, search_pattern, search_pattern, search_pattern, search_pattern, search_pattern])

        if status_filter:
            query += " AND status = ?"
            count_query += " AND status = ?"
            params.append(status_filter)
            count_params.append(status_filter)

        # Count total
        count_result = db.fetch_one(count_query, count_params)
        total_items = count_result["total"] if count_result else 0

        # Pagination & sorting
        offset = (page - 1) * page_size
        query += " ORDER BY created_at DESC LIMIT ? OFFSET ?"
        params.extend([page_size, offset])

        items = db.fetch_all(query, params)
        for item in items:
            item["balance"] = cls.get_nasabah_balance(item["id"])

        return items, total_items

    @classmethod
    def get_nasabah_by_id(cls, nasabah_id: str) -> Dict[str, Any]:
        """Get nasabah details by UUID."""
        nasabah = db.fetch_one("SELECT * FROM nasabah WHERE id = ?", (nasabah_id,))
        if not nasabah:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Nasabah tidak ditemukan.",
            )
        nasabah["balance"] = cls.get_nasabah_balance(nasabah_id)
        return nasabah

    @classmethod
    def update_nasabah(
        cls,
        nasabah_id: str,
        data: NasabahUpdateRequest,
        updater_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Update nasabah profile details and sync user login record."""
        existing = db.fetch_one("SELECT * FROM nasabah WHERE id = ?", (nasabah_id,))
        if not existing:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Nasabah tidak ditemukan.",
            )

        nik = data.nik if data.nik is not None else existing["nik"]
        if data.nik is not None and data.nik != existing["nik"]:
            dup_nik = db.fetch_one(
                "SELECT id FROM nasabah WHERE nik = ? AND id != ?",
                (data.nik, nasabah_id)
            )
            if dup_nik:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"NIK {data.nik} sudah terdaftar pada nasabah lain.",
                )

        name = data.name if data.name is not None else existing["name"]
        phone = data.phone if data.phone is not None else existing["phone"]
        address = data.address if data.address is not None else existing["address"]
        rt = data.rt if data.rt is not None else existing.get("rt")
        rw = data.rw if data.rw is not None else existing.get("rw")
        kelurahan = data.kelurahan if data.kelurahan is not None else existing.get("kelurahan")
        kecamatan = data.kecamatan if data.kecamatan is not None else existing.get("kecamatan")
        kabupaten_kota = data.kabupaten_kota if data.kabupaten_kota is not None else existing.get("kabupaten_kota")
        nasabah_category = data.nasabah_category if data.nasabah_category is not None else existing.get("nasabah_category", "Rumah Tangga/Individu")
        email = data.email if data.email is not None else existing["email"]

        with db.connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE nasabah
                SET nik = ?, name = ?, phone = ?, address = ?, rt = ?, rw = ?, kelurahan = ?, kecamatan = ?, kabupaten_kota = ?, nasabah_category = ?, email = ?, updated_by = ?, updated_at = datetime('now')
                WHERE id = ?
                """,
                (nik, name, phone, address, rt, rw, kelurahan, kecamatan, kabupaten_kota, nasabah_category, email, updater_id or "ADMIN", nasabah_id)
            )
            cursor.execute(
                """
                UPDATE users
                SET name = ?, phone = ?, email = ?, updated_by = ?, updated_at = datetime('now')
                WHERE nasabah_id = ?
                """,
                (name, phone, email, updater_id or "ADMIN", nasabah_id)
            )

        return cls.get_nasabah_by_id(nasabah_id)

    @classmethod
    def update_nasabah_status(
        cls,
        nasabah_id: str,
        status_val: str,
        updater_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Activate or inactivate a nasabah and their associated login account."""
        existing = db.fetch_one("SELECT * FROM nasabah WHERE id = ?", (nasabah_id,))
        if not existing:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Nasabah tidak ditemukan.",
            )

        is_active = 1 if status_val == "ACTIVE" else 0

        with db.connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE nasabah
                SET status = ?, updated_by = ?, updated_at = datetime('now')
                WHERE id = ?
                """,
                (status_val, updater_id or "ADMIN", nasabah_id)
            )
            cursor.execute(
                """
                UPDATE users
                SET status = ?, is_active = ?, updated_by = ?, updated_at = datetime('now')
                WHERE nasabah_id = ?
                """,
                (status_val, is_active, updater_id or "ADMIN", nasabah_id)
            )

        return cls.get_nasabah_by_id(nasabah_id)


nasabah_service = NasabahService()

File: backend/src/services/nasabahService.ts
import { v4 as uuidv4 } from 'uuid';
import { db, TxExecutor } from '../core/database.js';
import { hashPassword } from '../core/security.js';
import { AppError } from '../utils/response.js';
import {
  NasabahCreateInput,
  NasabahUpdateInput,
} from '../schemas/nasabah.schema.js';

export class NasabahService {
  public static async generateNextAccountNumber(tx?: TxExecutor): Promise<string> {
    const executor = tx || db;
    let row = await executor.fetchOne('SELECT last_number FROM account_sequences WHERE id = 1');
    let lastNumber = -1;
    if (!row) {
      await executor.execute('INSERT INTO account_sequences (id, last_number) VALUES (1, -1)');
    } else {
      lastNumber = Number(row.last_number);
    }

    const nextNumber = lastNumber + 1;
    if (nextNumber > 9999) {
      throw new AppError(
        'Kapasitas nomor rekening bsuvh0000-bsuvh9999 telah habis. Silakan hubungi pengembang sistem.',
        500,
        'ACCOUNT_CAPACITY_EXCEEDED'
      );
    }

    await executor.execute('UPDATE account_sequences SET last_number = ? WHERE id = 1', [
      nextNumber,
    ]);
    return `bsuvh${String(nextNumber).padStart(4, '0')}`;
  }

  public static async getNasabahBalance(nasabahId: string, tx?: TxExecutor): Promise<number> {
    const executor = tx || db;
    const balanceRow = await executor.fetchOne<any>(
      `SELECT COALESCE(SUM(credit) - SUM(debit), 0) as balance
       FROM transactions
       WHERE nasabah_id = ?`,
      [nasabahId]
    );
    return balanceRow ? Number(balanceRow.balance) : 0;
  }

  public static async createNasabah(
    data: NasabahCreateInput,
    registrationSource: 'SELF' | 'ADMIN' = 'SELF',
    creatorId: string | null = null
  ) {
    const existingNik = await db.fetchOne('SELECT id FROM nasabah WHERE nik = ?', [data.nik]);
    if (existingNik) {
      throw new AppError('NIK sudah terdaftar dalam sistem.', 400, 'NIK_EXISTS');
    }

    const nasabahId = uuidv4();
    const userId = uuidv4();
    const hashedPassword = await hashPassword(data.password);

    return await db.transaction(async (tx) => {
      const accountNo = await this.generateNextAccountNumber(tx);

      await tx.execute({
        sql: `INSERT INTO nasabah (
          id, customer_id, account_no, nik, name, phone, address,
          rt, rw, kelurahan, kecamatan, kabupaten_kota, nasabah_category, email,
          status, registration_source, created_by, updated_by
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'ACTIVE', ?, ?, ?)`,
        args: [
          nasabahId,
          accountNo,
          accountNo,
          data.nik,
          data.name,
          data.phone,
          data.address,
          data.rt || null,
          data.rw || null,
          data.kelurahan || null,
          data.kecamatan || null,
          data.kabupaten_kota || null,
          data.nasabah_category || 'Rumah Tangga/Individu',
          data.email || null,
          registrationSource,
          creatorId,
          creatorId,
        ],
      });

      await tx.execute({
        sql: `INSERT INTO users (
          id, username, name, email, phone, password_hash, role, nasabah_id, status, is_active, created_by
        ) VALUES (?, ?, ?, ?, ?, ?, 'NASABAH', ?, 'ACTIVE', 1, ?)`,
        args: [
          userId,
          accountNo,
          data.name,
          data.email || null,
          data.phone,
          hashedPassword,
          nasabahId,
          creatorId || 'SELF',
        ],
      });

      return {
        id: nasabahId,
        customer_id: accountNo,
        account_no: accountNo,
        nik: data.nik,
        name: data.name,
        phone: data.phone,
        address: data.address,
        rt: data.rt || null,
        rw: data.rw || null,
        kelurahan: data.kelurahan || null,
        kecamatan: data.kecamatan || null,
        kabupaten_kota: data.kabupaten_kota || null,
        nasabah_category: data.nasabah_category || 'Rumah Tangga/Individu',
        email: data.email || null,
        status: 'ACTIVE',
        registration_source: registrationSource,
        balance: 0,
      };
    });
  }

  public static async listNasabah(params: {
    search?: string;
    status?: string;
    category?: string;
    page?: number;
    page_size?: number;
  }) {
    const page = Math.max(1, Number(params.page) || 1);
    const pageSize = Math.max(1, Math.min(100, Number(params.page_size) || 10));
    const offset = (page - 1) * pageSize;

    const conditions: string[] = [];
    const args: any[] = [];

    if (params.search) {
      const s = `%${params.search.trim()}%`;
      conditions.push(
        '(name LIKE ? OR nik LIKE ? OR customer_id LIKE ? OR phone LIKE ? OR kelurahan LIKE ? OR kecamatan LIKE ?)'
      );
      args.push(s, s, s, s, s, s);
    }

    if (params.status) {
      conditions.push('status = ?');
      args.push(params.status);
    }

    if (params.category) {
      conditions.push('nasabah_category = ?');
      args.push(params.category);
    }

    const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    const countRow = await db.fetchOne<any>(
      `SELECT COUNT(*) as total FROM nasabah ${whereClause}`,
      args
    );
    const totalItems = countRow ? Number(countRow.total) : 0;
    const totalPages = Math.ceil(totalItems / pageSize);

    const rows = await db.fetchAll<any>(
      `SELECT * FROM nasabah ${whereClause} ORDER BY created_at DESC LIMIT ? OFFSET ?`,
      [...args, pageSize, offset]
    );

    // Calculate balances for each nasabah
    const items = await Promise.all(
      rows.map(async (row) => {
        const balance = await this.getNasabahBalance(row.id);
        return {
          ...row,
          balance,
        };
      })
    );

    return {
      items,
      pagination: {
        page,
        page_size: pageSize,
        total_items: totalItems,
        total_pages: totalPages,
      },
    };
  }

  public static async getNasabahById(id: string) {
    const nasabah = await db.fetchOne<any>(
      'SELECT * FROM nasabah WHERE id = ? OR customer_id = ? OR account_no = ? OR nik = ?',
      [id, id, id, id]
    );
    if (!nasabah) {
      throw new AppError('Data nasabah tidak ditemukan.', 404, 'NOT_FOUND');
    }
    const balance = await this.getNasabahBalance(nasabah.id);
    return {
      ...nasabah,
      balance,
    };
  }

  public static async updateNasabah(id: string, data: NasabahUpdateInput, updatedBy?: string) {
    const existing = await db.fetchOne<any>('SELECT * FROM nasabah WHERE id = ?', [id]);
    if (!existing) {
      throw new AppError('Data nasabah tidak ditemukan.', 404, 'NOT_FOUND');
    }

    if (data.nik && data.nik !== existing.nik) {
      const duplicateNik = await db.fetchOne('SELECT id FROM nasabah WHERE nik = ? AND id != ?', [
        data.nik,
        id,
      ]);
      if (duplicateNik) {
        throw new AppError('NIK sudah digunakan oleh nasabah lain.', 400, 'NIK_EXISTS');
      }
    }

    const updates: string[] = [];
    const args: any[] = [];

    const fields: (keyof NasabahUpdateInput)[] = [
      'nik',
      'name',
      'phone',
      'address',
      'rt',
      'rw',
      'kelurahan',
      'kecamatan',
      'kabupaten_kota',
      'nasabah_category',
      'email',
    ];

    for (const field of fields) {
      if (data[field] !== undefined) {
        updates.push(`${field} = ?`);
        args.push(data[field]);
      }
    }

    if (updates.length === 0) {
      return await this.getNasabahById(id);
    }

    updates.push("updated_at = datetime('now')");
    if (updatedBy) {
      updates.push('updated_by = ?');
      args.push(updatedBy);
    }

    args.push(id);
    await db.execute(`UPDATE nasabah SET ${updates.join(', ')} WHERE id = ?`, args);

    // Sync name and phone with associated user account if updated
    if (data.name || data.phone || data.email) {
      const userUpdates: string[] = [];
      const userArgs: any[] = [];
      if (data.name) {
        userUpdates.push('name = ?');
        userArgs.push(data.name);
      }
      if (data.phone) {
        userUpdates.push('phone = ?');
        userArgs.push(data.phone);
      }
      if (data.email !== undefined) {
        userUpdates.push('email = ?');
        userArgs.push(data.email);
      }
      if (userUpdates.length > 0) {
        userArgs.push(id);
        await db.execute(
          `UPDATE users SET ${userUpdates.join(', ')} WHERE nasabah_id = ?`,
          userArgs
        );
      }
    }

    return await this.getNasabahById(id);
  }

  public static async updateNasabahStatus(id: string, status: 'ACTIVE' | 'INACTIVE', updatedBy?: string) {
    const existing = await db.fetchOne<any>('SELECT * FROM nasabah WHERE id = ?', [id]);
    if (!existing) {
      throw new AppError('Data nasabah tidak ditemukan.', 404, 'NOT_FOUND');
    }

    await db.transaction(async (tx) => {
      await tx.execute({
        sql: `UPDATE nasabah SET status = ?, updated_at = datetime('now'), updated_by = ? WHERE id = ?`,
        args: [status, updatedBy || null, id],
      });

      const isActive = status === 'ACTIVE' ? 1 : 0;
      await tx.execute({
        sql: `UPDATE users SET status = ?, is_active = ?, updated_at = datetime('now') WHERE nasabah_id = ?`,
        args: [status, isActive, id],
      });
    });

    return await this.getNasabahById(id);
  }

  public static async getNasabahTransactions(
    nasabahId: string,
    params: { page?: number; page_size?: number; type?: string }
  ) {
    const page = Math.max(1, Number(params.page) || 1);
    const pageSize = Math.max(1, Math.min(100, Number(params.page_size) || 10));
    const offset = (page - 1) * pageSize;

    const conditions: string[] = ['t.nasabah_id = ?'];
    const args: any[] = [nasabahId];

    if (params.type) {
      conditions.push('t.type = ?');
      args.push(params.type);
    }

    const whereClause = `WHERE ${conditions.join(' AND ')}`;

    const countRow = await db.fetchOne<any>(
      `SELECT COUNT(*) as total FROM transactions t ${whereClause}`,
      args
    );
    const totalItems = countRow ? Number(countRow.total) : 0;
    const totalPages = Math.ceil(totalItems / pageSize);

    const rows = await db.fetchAll<any>(
      `SELECT t.*, 
              c.name as category_name,
              p.group_name as price_group_name,
              p.price_code as price_code
       FROM transactions t
       LEFT JOIN waste_categories c ON t.category_id = c.id
       LEFT JOIN waste_price_masters p ON t.price_id = p.id
       ${whereClause}
       ORDER BY t.transaction_date DESC, t.created_at DESC
       LIMIT ? OFFSET ?`,
      [...args, pageSize, offset]
    );

    const items = rows.map((row) => {
      let category = null;
      if (row.category_id || row.price_id) {
        const groupName = row.price_group_name;
        const catName = row.category_name || 'Sampah';
        const displayName = groupName ? `[${groupName}] ${catName}` : catName;
        category = {
          id: row.category_id || '',
          name: displayName,
          group_name: groupName,
          price_code: row.price_code,
        };
      }

      return {
        ...row,
        weight_kg: row.weight_gram !== null ? row.weight_gram / 1000.0 : null,
        category,
      };
    });

    return {
      items,
      pagination: {
        page,
        page_size: pageSize,
        total_items: totalItems,
        total_pages: totalPages,
      },
    };
  }
}

export const nasabahService = NasabahService;
